src/DotNet.Testcontainers/Services/ResourceReaper.cs

```diff
@@ -0,0 +1,219 @@
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper"/> has been started.</returns>
+    public static async Task<ResourceReaper> StartNew(Guid? sessionId = null)
+    {
+      var resourceReaper = new ResourceReaper(sessionId);
+
+      try
+      {
+        await resourceReaper.StartAsync().ConfigureAwait(false);
+      }
+      catch (Exception)
+      {
+        await resourceReaper.DisposeAsync().ConfigureAwait(false);
+        throw;
+      }
+
+      return resourceReaper;
+    }
+
+    /// <summary>
+    /// Returns the default <see cref="ResourceReaper"/> instance and starts it if necessary.
+    /// </summary>
+    /// <param name="ct">Cancellation token.</param>
+    /// <returns>Task that completes when the <see cref="ResourceReaper"/> has been started.</returns>
+    public static async Task<ResourceReaper> GetAndStartDefaultAsync()
+    {
+      if (defaultInstance != null && !defaultInstance.disposed)
+      {
+        return defaultInstance;
+      }
+
+      await defaultLock.WaitAsync().ConfigureAwait(false);
+
+      try
+      {
+        if (defaultInstance != null && !defaultInstance.disposed)
+        {
+          return defaultInstance;
+        }
+
+        defaultInstance = await StartNew(DefaultSessionId);
+
+        return defaultInstance;
+      }
+      finally
+      {
+        defaultLock.Release();
+      }
+    }
+
+    private async Task StartAsync()
+    {
+      await this.resourceReaperContainer.StartAsync().ConfigureAwait(false);
+
+      var ryukInitializedTaskSource = new TaskCompletionSource<bool>();
+
+#pragma warning disable 4014
+      Task.Run(() => this.MaintainRyukConnection(this.cancellationTokenSource.Token, ryukInitializedTaskSource));
+#pragma warning restore 4014
+
+      await ryukInitializedTaskSource.Task.ConfigureAwait(false);
+    }
+
+    private async Task MaintainRyukConnection(CancellationToken cancellationToken, TaskCompletionSource<bool> ryukInitializedTaskSource)
+    {
+      var initTimeoutTokenSrc = new CancellationTokenSource(TimeSpan.FromSeconds(10));
+      var hostname = this.resourceReaperContainer.Hostname;
+      var port = this.resourceReaperContainer.GetMappedPublicPort(RyukPort);
+
+      try
+      {
+        while (!cancellationToken.IsCancellationRequested && !initTimeoutTokenSrc.IsCancellationRequested)
+        {
+          try
+          {
+            using (var tcpClient = new TcpClient())
+            {
+              await tcpClient.ConnectAsync(hostname, port).ConfigureAwait(false);
+
+              var stream = tcpClient.GetStream();
+
+              if (!ryukInitializedTaskSource.Task.IsCompleted)
+              {
+                await this.SendSessionLabel(stream, cancellationToken).ConfigureAwait(false);
+                ryukInitializedTaskSource.SetResult(true);
+              }
+
+              while (!cancellationToken.IsCancellationRequested)
```

Comment:
@PSanetra is that loop necessary to keep the connection open?

Reply:
This is just in case if future ryuk decides to send more lines while the connection is open.